File: lib/python/src/aria/ops/timer.py
from __future__ import annotations

from logging import Logger
from time import time
from types import TracebackType
from typing import Iterable
from typing import List
from typing import Optional
from typing import Tuple
from typing import Type


class Timer:
    timers: List[Tuple[str, float, float]] = []

    def __init__(self, logger: Logger, name: str) -> None:
        self.logger = logger
        self.name = name

    def __enter__(self) -> Timer:
        self.logger.info(self.name)
        self.start_time = time()
        return self

    def __aenter__(self) -> Timer:
        self.logger.info(self.name)
        self.start_time = time()
        return self

    def __exit__(
        self,
        exc_type: Optional[Type[BaseException]] = None,
        exc_value: Optional[BaseException] = None,
        traceback: Optional[TracebackType] = None,
    ) -> None:
        end_time = time()
        self.timers.append((self.name, self.start_time, end_time))
        self.logger.info(f"finished {self.name} in {end_time - self.start_time:.2}s")

    def __aexit__(
        self,
        exc_type: Optional[Type[BaseException]] = None,
        exc_value: Optional[BaseException] = None,
        traceback: Optional[TracebackType] = None,
    ) -> None:
        end_time = time()
        self.timers.append((self.name, self.start_time, end_time))
        self.logger.info(f"finished {self.name} in {end_time - self.start_time:.2}s")

    @classmethod
    def graph(cls) -> str:
        headers = ["Operation", "Time"]
        full_headers = []
        width = 88
        time_min = time()
        time_max = 0.0
        name_max = len(headers[0])
        for name, start, end in cls.timers:
            if start < time_min:
                time_min = start
            if end > time_max:
                time_max = end
            if len(name) > name_max:
                name_max = len(name)

        graph_width = width - name_max - 10
        step = (time_max - time_min) / graph_width
        headers.append("t=0s")
        headers.append(f"t={_to_time(time_max - time_min)}")

        full_headers.append(f"{headers[0]:<{name_max}}")
        full_headers.append(f"{headers[1]:<8}")
        full_headers.append(
            f"{headers[2]:<{int(graph_width / 2)}}{headers[3]:>{int((graph_width + 1) / 2)}}"
        )

        sorted_intervals: Iterable[Tuple[str, float, float]] = sorted(
            cls.timers, key=lambda i: i[1]
        )
        graph = "Timing Graph: \n"
        graph += _hline(full_headers, "U")
        graph += "│".join(full_headers) + "\n"
        graph += _hline(full_headers)
        for name, start, end in sorted_intervals:
            graph += (
                f"{name:<{name_max}}│"
                f"{_to_time(end - start):<8}│"
                + _graph_timespan(
                    (start - time_min) / step, (end - time_min) / step, graph_width
                )
                + "\n"
            )
        graph += _hline(full_headers, "L")
        return graph


def _hline(headers: List, loc: str = "M") -> str:
    hl = ""
    for header in headers:
        for x in range(len(header)):
            hl += "━"
        if loc == "U":
            hl += "┯"
        elif loc == "M":
            hl += "┿"
        else:  # loc = L
            hl += "┷"
    return f"{hl[:-1]}\n"


def _graph_timespan(start: float, end: float, graph_width: int) -> str:
    line = ""

    def in_range(x_coord: float) -> bool:
        return start <= x_coord < end

    for x in range(graph_width):
        if in_range(x - 1) and in_range(x) and in_range(x + 1):
            line += "─"
        elif in_range(x - 1) and in_range(x):
            line += "╼"
        elif in_range(x) and in_range(x + 1):
            line += "╾"
        elif in_range(x):
            line += "━"
        else:
            line += " "
    return f"{line}"


def _to_time(seconds: float) -> str:
    if seconds > 3600:
        hours = seconds / 3600
        return f"{hours:.2f}h"
    elif seconds > 120:
        minutes = seconds / 60
        return f"{minutes:.2f}m"
    return f"{seconds:.2f}s"

File: lib/java/AdapterLibrary/src/main/kotlin/com/vmware/aria/operations/Timer.kt
/*
 * Copyright 2023 VMware, Inc.
 * SPDX-License-Identifier: Apache-2.0
 */
@file:JvmName("Timing")

package com.vmware.aria.operations

import com.vmware.aria.operations.Timer.Companion.timers
import kotlin.math.roundToInt
import kotlin.time.ComparableTimeMark
import kotlin.time.Duration
import kotlin.time.DurationUnit
import kotlin.time.TimeSource


/**
 * Creates and starts a new timer. The timer will run until [stop] is called.
 * Timers are automatically registered and can be graphed in aggregate using the [graph] method.
 * @property name A name to identify the timer.
 */
class Timer constructor(val name: String) {
    internal val startTime: ComparableTimeMark
    internal var endTime: ComparableTimeMark? = null
    internal val duration: Duration
        get() = endTime?.let { it - startTime } ?: Duration.INFINITE

    init {
        startTime = TimeSource.Monotonic.markNow()
        timers.add(this)
    }

    /**
     * Stops the timer. Once stopped, a timer cannot be restarted.
     */
    fun stop() {
        if (endTime == null) {
            endTime = TimeSource.Monotonic.markNow()
        }
    }

    /**
     * Gets the duration of the timer in milliseconds
     */
    fun getDuration(): Long =
        duration.inWholeMilliseconds

    /**
     * @return The timer's name and duration. If the timer has not been stopped, the duration
     * will read as 'Infinity'
     */
    override fun toString(): String =
        "$name: ${duration.toTime()}"

    companion object {
        internal val timers: MutableList<Timer> = mutableListOf()

    }
}

/**
 * Times the function / closure. The time is not returned but can be graphed using the
 * [graph] function.
 *
 * @param name The name for this timer
 * @param fn The function or closure to time
 * @return The result of [fn]
 */
fun <T> time(name: String, fn: () -> T?): T? {
    val timer = Timer(name)
    val returnValue = fn()
    timer.stop()
    return returnValue
}

private enum class LineStyle(val horizontalChar: String, val junctionChar: String) {
    UPPER("━", "┯"),
    MIDDLE("━", "┿"),
    LOWER("━", "┷"),
}

/**
 * Graphs all the [Timers][Timer] that have been started since program start, or the last
 * invocation of [clearTimers].
 */
fun graph(): String {
    val width = 88
    val sortedTimers = timers.sortedBy { timer -> timer.startTime }
    val timeMin = sortedTimers.first().startTime
    val timeMax =
        timers.maxOf { timer -> timer.endTime ?: TimeSource.Monotonic.markNow() }
    val nameMax = timers.maxOf { timer -> timer.name.length }
    val graphWidth = width - nameMax - 10
    val duration = timeMax - timeMin
    val headers = listOf(
        "Operation".padEnd(nameMax),
        "Time".padEnd(8),
        "t=0s".padEnd(graphWidth / 2) +
                "t=${duration.toTime()}".padStart((graphWidth + 1) / 2)
    )
    val graph = StringBuilder()
    graph.appendLine("Timing Graph: ")
    graph.appendLine(horizontalLine(headers, LineStyle.UPPER))
    graph.appendLine(headers.joinToString("│"))
    graph.appendLine(horizontalLine(headers, LineStyle.MIDDLE))
    for (timer in sortedTimers) {
        graph.appendLine(display(timer, timeMin, timeMax, headers))
    }
    graph.appendLine(horizontalLine(headers, LineStyle.LOWER))
    return graph.toString()
}

/**
 * Clears all currently running and stopped timers. Use this to start a new timing graph.
 */
fun clearTimers() {
    timers.clear()
}

private fun horizontalLine(headers: List<String>, style: LineStyle): String =
    headers.joinToString("") { header ->
        style.horizontalChar.repeat(header.length) + style.junctionChar
    }.dropLast(1)

private fun display(
    timer: Timer,
    startTime: ComparableTimeMark,
    endTime: ComparableTimeMark,
    headers: List<String>,
): String =
    listOf(
        timer.name.padEnd(headers[0].length),
        timer.duration.toTime().padEnd(headers[1].length),
        graphTimer(timer, startTime, endTime, headers[2].length)
    ).joinToString("│")

private fun graphTimer(
    timer: Timer,
    startTime: ComparableTimeMark,
    endTime: ComparableTimeMark,
    width: Int,
): String {
    val step = (endTime - startTime).div(width - 1)
    val start = (timer.startTime - startTime).div(step).roundToInt()
    val end = ((timer.endTime ?: endTime) - startTime).div(step).roundToInt()
    var line = "".padEnd(start)
    if (start == end) {
        line += "━"
    } else {
        line += "╾"
        line += "─".repeat(end - start - 1)
        line += timer.endTime?.let { "╼" } ?: "─"
    }
    return line
}

private fun Duration.toTime(): String {
    val seconds = toDouble(DurationUnit.SECONDS)
    if (seconds >= 3600)
        return toString(DurationUnit.HOURS, 2)
    else if (seconds > 120)
        return toString(DurationUnit.MINUTES, 2)
    return toString(DurationUnit.SECONDS, 2)
}
